src/core/MIGRATION.md

```diff
@@ -653,20 +671,66 @@ At the very least, any plugin exposing an extension point should do so with firs
-* The `x-pack` directory moved into `node_modules` at build time, so a relative import from `x-pack` to `src` that worked during development would break once a Kibana distribution was built. This is still a problem today, but the fix is in flight via issue [#32722](https://github.com/elastic/kibana/pull/32722).
+Examples of code that could be shared statically:
+- Constants. Strings and numbers that do not ever change (even between Kibana versions)
+- React components. Make sure these components are not dependent on or pre-wired to Core services. In many of these cases you can export a HOC that takes the Core service and returns a component wired up to that particular service instance.
+- Pure computation functions.
+
+Examples of code that could **not** be shared statically:
+- A function that queries Elasticsearch, but does not take an ES Client as a parameter.
+- A React component that reads `uiSettings` from a variable defined in the module.
+- A function that allows other plugins to register values that get pushed into an array defined internally to the module.
+
+#### So what happens if I don't listen to you?
+
+One of the outcomes of producing a stable core API in Kibana is that a single plugin version should work across all installations of the same major version of Kibana. To enable this, we are planning to use "immutable bundles" that are built once, and compatible with every minor version of that Kibana major version.
+
+- Any code you export to other plugins will get copied into their bundles, so if a plugin is built for 8.1 and is running on Kibana 8.2, any modules it imported that changed will not be updated in that plugin.
+- When a plugin is disabled, its static exports will still be importable by any other plugin. This can result in undesirable effects where it _appears_ like a plugin is enabled when it is not. In the worst case, it can result in an unexpected user experience where features that should have been disabled are not.
+- Exported modules will be copied into plugins that import them. Any state your plugin populates to a exported module will not be the same instance of the module that dependent plugins are using. Effectively, impure exported modules simply won't work and your plugin will not have access to the copy of the module in other plugins.
+
+#### Concrete Example
+
+Ok, you've decided you want to export static code from your plugin, how do you do it? In principle, the New Platform only considers values exported from `my_plugin/public` and `my_plugin/server` to be stable. Our tooling will enforce that these APIs do not break between major versions (in the future). All code shared among plugins should be exported in these modules like so:
+
+```ts
+// my_plugin/public/index.ts
+export { MyPureComponent } from './components';
```

Comment:
I'm down for this API. Although, worth noticing that we mix static exports for other plugins and exports for Kibana core in one file.

Reply:
I think we may have to consider new additions to the export contract for Core as breaking changes, or at the very least implement it a non-breaking way.

Of course this is still trick if we add _optional_ exports to the PluginModule contract. Core should be able to detect if the export _seems_ to follow the shape it's expecting, though that's not perfect. We do this already with the `plugin` export by ensuring that the plugin it instantiates has the lifecycle methods.

Reply:
>  or at the very least implement it a non-breaking way.

How we can be sure for community-plugins? 

>  as breaking changes

that can be a problem  with out long release cycle, if we follow semver ofc.

Reply:
> How we can be sure for community-plugins?

Reply:
The "simplist" way to add new exports in a non-breaking is make new exports that are in that module contract optional, so we only use that export if it's present.

Of course this could still clash with a name that a 3rd party developer chooses to export. We could change it so that extra exports from a plugin are in a sub-object called `extras`?

```js
export const extras = {
  MyComponent,
  myUtilFunc,
};

export const plugin = () => { ... }
```

Reply:
> We could change it so that extra exports from a plugin are in a sub-object called extras?

I'd vote for reserving a new file for this type of exports, e.g. `my_plugin/public/exports.ts`, `my_plugin/public/static_exports.ts`, or `my_plugin/public/extras.ts`. It makes it easy to discern whether a plugin has static exports or not by scanning the directory, it will scale better if users have large plugin definitions and/or many static exports, and it looks more conventional from the consumer's standpoint IMO:

```js
import { MyPureComponent } from '../my_plugin/public/static_exports';
// vs
import { extras: { MyPureComponent } } from '../my_plugin/public';
```

Reply:
> It makes it easy to discern whether a plugin has static exports or not by scanning the directory, it will scale better if users have large plugin definitions and/or many static exports, and it looks more conventional from the consumer's standpoint IMO

I actually disagree, I think it's more discoverable if the exports are in a single place and auto-complete/intellisense can give you a list of what's available rather than having to trial and error for an `my_plugin/public/extras` directory that may or may not exist.

There's also a technical problem with having multiple files serve as the "public API" to the module which is that we're moving to an "immutable bundle" pattern which should allow you to develop against other plugins without needing their entire source code tree, just the built bundle (and a index.d.ts file for types). If we had a separate import path for static imports, we'd have to build 2 bundles so that we could import from `extras`.

@elastic/kibana-platform We need to get consensus on this. I'm leaning towards the proposal in this document, that is plugin authors can export whatever they'd like from `my_plugin/public` and `my_plugin/server`. This does put us at risk for more breaking changes, however I'm having a hard time contriving any more examples of what Core might need to import from this top-level, other than `plugin` and `config` (which is only the server-side).

Reply:
I think your technical counterpoint is sound, but in terms of

> I actually disagree, I think it's more discoverable if the exports are in a single place and auto-complete/intellisense can give you a list of what's available rather than having to trial and error for an my_plugin/public/extras directory that may or may not exist.

I would just caution that this isn't how everyone develops. For example, it's not how I work. :) So for the sake of inclusivity, I'd encourage us all to think in terms of structuring our code for clarity using the primitives exposed to all of us (names, files, and folders), and using intellisense as a nice enhancement but not a requirement.